[DTensor][TP][Random] Introduce TensorParallelRNGTracker to integrate parallel RNG state with Tensor Parallel

XilunWu opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #104215
* #104440
* __->__ #103910
* #103235

This PR enables the automatic use of `TensorParallelRNGTracker` in Tensor Parallel api. Some unit tests are going to be added to cover.

XilunWu committed:
[DTensor][TP][Random] Introduce TensorParallelRNGTracker to integrate parallel RNG state with Tensor Parallel

[ghstack-poisoned]

XilunWu committed:
Update on "[DTensor][TP][Random] Introduce TensorParallelRNGTracker to integrate parallel RNG state with Tensor Parallel"




[ghstack-poisoned]

XilunWu committed:
Update on "[DTensor][TP][Random] Introduce TensorParallelRNGTracker to integrate parallel RNG state with Tensor Parallel"


This PR enables the automatic use of `TensorParallelRNGTracker` in Tensor Parallel api. Some unit tests are going to be added to cover.

[ghstack-poisoned]

XilunWu committed:
Update on "[DTensor][TP][Random] Introduce TensorParallelRNGTracker to integrate parallel RNG state with Tensor Parallel"


This PR enables the automatic use of `TensorParallelRNGTracker` in Tensor Parallel api. Some unit tests are going to be added to cover.

[ghstack-poisoned]

fduwjj commented on torch/distributed/tensor/parallel/api.py:
```diff
@@ -82,6 +87,14 @@ def parallelize_module(  # type: ignore[return]
+        not isinstance(random._rng_tracker, TensorParallelRNGTracker)
+    ):
+        random._rng_tracker = TensorParallelRNGTracker()
+        random._rng_tracker._manual_seed(device_mesh, base_seed=(1234 + 2718), tp_dim=tp_mesh_dim)
```
why 1234 + 2718?

XilunWu replied:
`1234` is the default seed used and `2718` is the "magic number" added to tensor parallel seed in Megatron. I will move `2718` back to TP RNG tracker and leave `1234` here.

fduwjj commented on torch/distributed/tensor/parallel/api.py:
```diff
@@ -82,6 +87,14 @@ def parallelize_module(  # type: ignore[return]
+    # instantiate a TP RNG state tracker if it's not there
+    if (
+        is_rng_supported_mesh(device_mesh) and
+        not isinstance(random._rng_tracker, TensorParallelRNGTracker)
```
Can we also pass in the Tracker in case users want to use OffsetBasedRNGTracker?

XilunWu replied:
then we need a way to control it (e.g. a switch to decide which tracker to use for TP). Maybe a config variable.

fduwjj replied:
I mean passing in as a input variable of the API.

fduwjj commented on torch/distributed/_tensor/random.py:
```diff
@@ -297,4 +299,42 @@ def _calc_shard_linear_idx(
         return shard_linear_idx
 
 
+class TensorParallelRNGTracker(CudaRNGStateTracker):
```
Can we add a unit test for this one? Thanks!

XilunWu committed:
Update on "[DTensor][TP][Random] Introduce TensorParallelRNGTracker to integrate parallel RNG state with Tensor Parallel"


This PR enables the automatic use of `TensorParallelRNGTracker` in Tensor Parallel api. Some unit tests are going to be added to cover.

[ghstack-poisoned]

XilunWu committed:
Update on "[DTensor][TP][Random] Introduce TensorParallelRNGTracker to integrate parallel RNG state with Tensor Parallel"


This PR enables the automatic use of `TensorParallelRNGTracker` in Tensor Parallel api. Some unit tests are going to be added to cover.

[ghstack-poisoned]

XilunWu commented on torch/distributed/tensor/parallel/api.py:
```diff
@@ -37,6 +44,8 @@ def parallelize_module(  # type: ignore[return]
     device_mesh: DeviceMesh,
     parallelize_plan: Union[ParallelStyle, Dict[str, ParallelStyle]],
     tp_mesh_dim: int = 0,
+    enable_distribute_region: bool = True,
```
"enable_distribute_region" -> "sync_rng_state"

XilunWu commented on torch/distributed/tensor/parallel/api.py:
```diff
@@ -37,6 +44,8 @@ def parallelize_module(  # type: ignore[return]
     parallelize_plan: Union[ParallelStyle, Dict[str, ParallelStyle]],
     tp_mesh_dim: int = 0,
+    enable_distribute_region: bool = True,
+    rng_tracker_type: Type[CudaRNGStateTracker] = OffsetBasedRNGTracker,
```
leave this to another PR

XilunWu committed:
Update on "[DTensor][TP][Random] Introduce TensorParallelRNGTracker to integrate parallel RNG state with Tensor Parallel"


This PR enables the automatic use of `TensorParallelRNGTracker` in Tensor Parallel api. Some unit tests are going to be added to cover.

[ghstack-poisoned]

XilunWu committed:
Update on "[DTensor][TP][Random] Introduce TensorParallelRNGTracker to integrate parallel RNG state with Tensor Parallel"


This PR enables the automatic use of `TensorParallelRNGTracker` in Tensor Parallel api. Some unit tests are going to be added to cover.

[ghstack-poisoned]

wanchaol commented on test/distributed/tensor/parallel/test_tp_random_state.py:
```diff
@@ -0,0 +1,185 @@
+            assert isinstance(out, torch.Tensor)
+            tensor_local = out
+            tensor_gather = device_mesh.all_gather(tensor_local, mesh_dim=1, gather_dim=0)
+            for other_rank in range(device_mesh.size(dim=1)):
```
I am wondering if it's possible to make the checking logic be a util function instead? I see both the model init test case and the dropout test case uses similar logic, and we used it for like three times, so we should make this a util instead to avoid redundant code

wanchaol reviewed (commented):
Nice work! mostly looks good to me, I have some suggestions on the testing and the default v alue of the additional arg, see comments inlined.

XilunWu commented on test/distributed/tensor/parallel/test_tp_random_state.py:
```diff
@@ -0,0 +1,185 @@
+        #   2. residual dropout which belongs to a non-tensor-parallel region
+        mesh = torch.arange(self.world_size).reshape(2, 2)
+        device_mesh = DeviceMesh(self.device_type, mesh)
+        tp_rank = device_mesh.get_coordinate()[0]  # the tensor parallel dimension is 0
```
[parallelize_module](https://github.com/pytorch/pytorch/blob/main/torch/distributed/tensor/parallel/api.py#L35-L40) has the default `tp_mesh_dim` value 0. I tried printing the 1-d mesh created and the result is:
device_mesh: [[0, 1], [2, 3]]
1d_mesh created in TP api: [0, 2] and [1, 3]

wanchaol replied:
Oh I think for the case of 2-D parallel what we are doing now is mainly tp_dim = 1, so user might need to specify their tp_dim when combining with data parallel. I think the default argument specified as 0 because there might be case we only want to perform tensor parallel without data parallel, in this case it's a 1-D mesh.

But I guess for the case of checking the content whether tp is dim 0 or dim 1 does not matter that much

XilunWu replied:
I  agree. For unit test this shouldn't bother.

XilunWu committed:
Update on "[DTensor][TP][Random] Introduce TensorParallelRNGTracker to integrate parallel RNG state with Tensor Parallel"


This PR enables the automatic use of `TensorParallelRNGTracker` in Tensor Parallel api. Some unit tests are going to be added to cover.

[ghstack-poisoned]

wanchaol reviewed (approved):
lgtm, thanks!

fduwjj commented on torch/distributed/tensor/parallel/api.py:
```diff
@@ -94,6 +116,15 @@ def parallelize_module(  # type: ignore[return]
             return _parallelize_multihead_attn(module, device_mesh)
         elif _is_mlp_for_pairwise_parallel(module):
             return _parallelize_mlp(module, device_mesh, parallelize_plan)
+        elif isinstance(module, torch.nn.Dropout):
+            distribute_module(
+                module,
+                device_mesh,
+                None,
+                input_fn=parallelize_plan._prepare_input,  # type: ignore[arg-type, misc] # pyre-ignore[6]
+                output_fn=parallelize_plan._prepare_output,  # type: ignore[arg-type, misc] # pyre-ignore[6]
+            )
+            return module
```
This part I think is a little bit intrusive and a temporary hack. Here is My two cents also mentioned in another PR:

Let's make this optional for users or make this as an extension which user can pass in because what if today users all use pure DTensor ops which means this logic is not needed at all or what if we want to have other special handling ops like this? (For example, softmax, layernorm, etc.)

If this is just for benchmark purpose, can we exclude this part outside this PR?

XilunWu replied:
Yes I can move dropout part into another PR. I think if we have a solution in TP, it would be great: currently we can only parallelize `mlp`, `linear`, and `mha`. User can not pass in other layers in `parallelize_plan` argument and parallelize it, if I understand correctly. I think we need to have some solution for this case.

XilunWu replied:
moved this to #104215

fduwjj commented on torch/distributed/_tensor/random.py:
```diff
@@ -297,3 +299,43 @@ def _calc_shard_linear_idx(
+        coordinate = device_mesh.get_coordinate()
+        assert coordinate is not None
+        tensor_parallel_rank = coordinate[tp_dim]
+        # this magic number 2718 comes from Megatron's code
+        # (https://github.com/NVIDIA/Megatron-LM/blob/main/megatron/core/tensor_parallel/random.py)
+        tensor_parallel_seed = base_seed + 2718 + tensor_parallel_rank
```
Opps, sorry, I thought the comments are out but it turns out they are not... My bad.

XilunWu commented on torch/distributed/_tensor/random.py:
```diff
@@ -297,3 +299,43 @@ def _calc_shard_linear_idx(
+
+        if self.distribute_region_enabled:
+            with torch.random.fork_rng(self._devices):
+                torch.cuda.set_rng_state(self.rng_states["tensor-parallel-rng"])
+                try:
+                    yield
+                finally:
+                    self.rng_states["tensor-parallel-rng"] = torch.cuda.get_rng_state()
```
Yes it will be the same unless user calls `manual_seed` to change it. This matches with megatron's setup.

XilunWu replied:
Several things:

- in every run of model training, Megatron uses either a user pass-in random seed from `config` or 1234 as the base random seed.
- The base random seed is used as the Data Parallel seed, if user of Megatron wants to have same data loading across DP groups, or base random seed plus DP group rank (e.g. DP group 2 has random seed 1234+2) if different data loading across DP groups is wanted. This seed is used for non-tensor-parallel region (e.g. data loading, reidual dropout in self-attention).
- Then, the Tensor Parallel seed equals DP seed + 2718 (the magic number) + TP rank so each rank within a TP group will have different random seed for tensor-parallel region random ops.

XilunWu commented on torch/distributed/tensor/parallel/api.py:
```diff
@@ -82,6 +87,14 @@ def parallelize_module(  # type: ignore[return]
+        is_rng_supported_mesh(device_mesh) and
+        not isinstance(random._rng_tracker, TensorParallelRNGTracker)
+    ):
+        random._rng_tracker = TensorParallelRNGTracker()
```
let's put that in another PR maybe, because I'm not quite sure how this argument should be used. Is it an optional RNG tracker argument, and we use the default value when it's None? But we don't want user to know RNG tracker, right? That's kinda unknown to us for now.

XilunWu committed:
Update on "[DTensor][TP][Random] Introduce TensorParallelRNGTracker to integrate parallel RNG state with Tensor Parallel"


This PR enables the automatic use of `TensorParallelRNGTracker` in Tensor Parallel api. Some unit tests are going to be added to cover.

[ghstack-poisoned]

XilunWu commented on torch/distributed/tensor/parallel/api.py:
```diff
@@ -62,6 +68,12 @@ def parallelize_module(  # type: ignore[return]
         tp_mesh_dim (int):
             The dimension of ``device_mesh`` where we perform
             Tensor Parallelism on.
+        sync_rng_state (bool):
+            This option represents whether or not the parallelized module uses the parallel
+            Random Number Generator to perform randomized tensor operators (e.g. dropout).
+            DTensor is responsible for maintaining the parallel RNG in a synchronous state
+            across ranks.
+            Default: False
```
They are orthogonal:

- random ops will use `_rng_tracker.distribute_region` to perform the computation. It doesn't know how `_rng_tracker` uses RNG state.
- `TensorParallelRNGTracker` is the type of RNG tracker to use in Tensor Parallel which mimics Megatron's use of random seeds: different seed within a TP group for tensor-parallel region, same seed for non-tensor-parallel region (e.g. data loading, residual dropout).

XilunWu replied:
After talking to @fduwjj offline, we decide to move this argument out of this PR and put it into a separate PR along with the option argument rng_tracker_style.

XilunWu committed:
Update on "[DTensor][TP][Random] Introduce TensorParallelRNGTracker to integrate parallel RNG state with Tensor Parallel"


This PR enables the automatic use of `TensorParallelRNGTracker` in Tensor Parallel api. Some unit tests are going to be added to cover.

[ghstack-poisoned]

fduwjj reviewed (approved):
thanks!

XilunWu commented:
regarding the RNG tracker for TP, we do and plan to do:

1. make TP RNG tracker the only approach that TP uses RNG. But we are open to offer a API to swtich if user requests in future.
2. we should offer a API under TP to allow user to set up random seed. either by a function call or through a config argument.

XilunWu commented:
@pytorchbot merge

XilunWu commented:
@pytorchmergebot merge
